module/validation/receipt_validator_test.go

```diff
@@ -850,3 +925,222 @@ func (s *ReceiptValidationSuite) TestValidateReceiptResultHasEnoughReceipts() {
+//   - per API contract it should return a `module.UnknownBlockError`
+//   - should _not_ be misinterpreted as an invalid receipt, i.e. should not receive an `engine.InvalidInputError`
+func (s *ReceiptValidationSuite) TestReceiptNoBlock() {
+	s.publicKey.On("Verify", mock.Anything, mock.Anything, mock.Anything).Return(true, nil).Maybe()
+
+	// Initially, s.LatestExecutionResult points to the result for s.LatestSealedBlock. We construct the chain:
+	//   LatestSealedBlock <-- unknownExecutedBlock  <-- candidate(r)
+	// where `r` denotes an execution receipt for block `unknownExecutedBlock`
+	unknownExecutedBlock := unittest.BlockWithParentFixture(s.LatestSealedBlock.Header)
+	r := unittest.ExecutionReceiptFixture(
+		unittest.WithExecutorID(s.ExeID), // valid executor
+		unittest.WithResult(unittest.ExecutionResultFixture(
+			unittest.WithBlock(unknownExecutedBlock),
+			unittest.WithPreviousResult(*s.LatestExecutionResult)), // known parent result
+		)) // but the ID of the executed block is randomly chosen, i.e. unknown
+
+	// attempting to validate receipt `r` should fail with an `module.UnknownBlockError`
+	err := s.receiptValidator.Validate(r)
+	s.Require().Error(err, "should reject invalid receipt")
+	s.Assert().True(module.IsUnknownBlockError(err), err)
+	s.Assert().False(engine.IsInvalidInputError(err), err)
+
+	// attempting to validate a block, whose payload contains receipt `r` should fail with an `module.UnknownBlockError`
+	candidate := unittest.BlockWithParentFixture(unknownExecutedBlock.Header)
+	candidate.SetPayload(unittest.PayloadFixture(unittest.WithReceipts(r)))
+	err = s.receiptValidator.ValidatePayload(candidate)
+	s.Require().Error(err, "should reject invalid receipt")
+	s.Assert().True(module.IsUnknownBlockError(err), err)
+	s.Assert().False(engine.IsInvalidInputError(err), err)
+}
+
+// TestException_SealsHighestInFork tests that unexpected exceptions raised by the dependency
+// `receiptValidator.headers.ByBlockID(..)` are escalated and not misinterpreted as
+// `InvalidInputError` or `UnknownBlockError` or `UnknownResultError`
+func (s *ReceiptValidationSuite) TestException_HeadersByBlockID() {
+	s.publicKey.On("Verify", mock.Anything, mock.Anything, mock.Anything).Return(true, nil).Maybe()
+
+	//s.T().Run("seals yields exception on Identity retrieval", func(t *testing.T) {
+	valSubgrph := s.ValidSubgraphFixture()
+	s.AddSubgraphFixtureToMempools(valSubgrph)
+
+	receipt := unittest.ExecutionReceiptFixture(unittest.WithExecutorID(s.ExeID), unittest.WithResult(valSubgrph.Result))
+	candidate := unittest.BlockWithParentFixture(valSubgrph.Block.Header)
+	candidate.SetPayload(unittest.PayloadFixture(unittest.WithReceipts(receipt)))
+
+	// receiptValidator.headers yields exception on retrieving any block header
+	*s.HeadersDB = mock_storage.Headers{} // receiptValidator has pointer to this field, which we override with a new state mock
```

Comment:
```suggestion
	*s.HeadersDB = *mock_storage.NewHeaders(s.T()) // receiptValidator has pointer to this field, which we override with a new state mock
```
Would advice to use `NewHeaders` to make sure that call will be indeed executed.

Reply:
👍 done in https://github.com/onflow/flow-go/pull/5211/commits/7623a680e8ca41258f7451aaa6db229da2efa935